### DISCUSSION TITLE
Create a standard for `.astro` file exports

### DISCUSSION THREAD

**Discussion by bholmesdev** [withastro/roadmap#592] [Proposal]
### Body

# Summary

Introduce a standard for "exports" from an Astro file, confirming how future magic exports in Astro (and potentially user exports) should be done.

# Background & Motivation

`.astro` files have an incomplete story for `export`ing types and arbitrary JS from frontmatter blocks. We have the following conventions:
- two exports are "reserved" by Astro: `prerender` and `getStaticPaths()`. Each have special rules detailed below.
- `export type` from any Astro file is fully supported.
- `export [arbitrary JS]` _can_ work, though this is not documented / encouraged.

For reserved exports, Astro core and the compiler are clever about "hoisting" these exports out of the enclosing scope. This allows for static analysis of exports without rendering the entire Astro file. This is colocation of exports in frontmatter is convenient... at the cost of closure footguns. This rough edge has existed in `getStaticPaths()` since pre-1.0:

```astro
---
// Example 1: works fine
import { getCollection } from 'astro:content';

export async function getStaticPaths() {
  const collection = await getCollection('blog');
  return collection.map(...);
}

// Example 2: fails!
import { getCollection } from 'astro:content';

const collection = await getCollection('blog');

export async function getStaticPaths() {
  return collection.map(...);
}
---
```

If we try to build the second example, we'll get the following error:

```
 error   posts is not defined
ReferenceError: posts is not defined
    at Module.getStaticPaths
```

This is because, when hoisting for static analysis, **outside variables cannot be referenced.** The core team has sited this limitation as a reason to _avoid_ magic exports for new features. A recent exploration It has also proven a blocker to letting users `export` from an Astro file, which sets Astro apart from frameworks like Svelte or Vue (which [each feature a `module` convention](https://svelte.dev/tutorial/module-exports)).

# Goals

- Confirm how future magic exports in Astro (and potentially user exports) should be done.
- Nice to have: Create a convention for `exports` that can solve the `getStaticPaths()` closure problem.

**Comment by bholmesdev** [comment]
## Idea 1: `---exports` block

One option could be a new frontmatter block above our standard fence. This could be denoted with `---exports`, inspired by [Svelte's `content="module"` convention](https://svelte.dev/tutorial/module-exports):

```ts
---exports
export async function getStaticPaths() {
  const collection = await getCollection('blog');
  return collection.map(...);
}
---
const entry = Astro.props;
---
<h1>{entry.data.title}</h1>
```

At the compiler level, `---exports` would be split to a separate output chunk we could statically analyze. This chunk would not inject Astro's `$$result` or other component rendering imports.

Pros: keeps colocation of exports, solves hoisting problem
Cons: requires compiler and language tooling changes, could be confusing to read

Open questions:
- Would the `Astro` global be available from this scope?
- Is there a non-breaking migration path?

## Idea 2: enforce separate files

Another option could be... separate files. I know, blasphemy! Though SFC frameworks like Svelte have enforced a separate file of the same name for reserved export like `load` and `action`:
```bash
app/
  +page.ts # data loaders
  +page.svelte # component renderer, received data as props
```

Astro could enforce the same standard, with `getStaticPaths()` falling in a separate `[route-name].ts` file alongside your `[route-name].astro`:

```bash
pages/blog/
  [slug].astro
  [slug].ts
```

Pros: no compiler or language tool changes, solves hoisting problem
Cons: more files = bad to some users

**Reply by lilnasy** [reply]
I hope there's a better option than `---exports`. When it follows the per-request frontmatter, it  could be construed as the name of the boundary rather than the block that follows it.
```astro
---
const entry = Astro.props;
---exports
export async function getStaticPaths() {
  const collection = await getCollection('blog');
  return collection.map(...);
}
---
<h1>{entry.data.title}</h1>
```

Another idea is using labels to name a block that can't use the scope, and exists outside the per-request lifecycle of the rest of the frontmatter.
```astro
---
import { getCollection } from 'astro:content'
const collection = await getCollection('blog')

metadata: {
    const prerender = true
    
    function getStaticPaths() {
        ...
    }
}
---
```

**Comment by lilnasy** [comment]
I'm hoping this includes a way to expose exports to a middleware. That will allow, for example, a page to export `revalidate = 60` and a caching middleware to cache the response for 60 seconds.
